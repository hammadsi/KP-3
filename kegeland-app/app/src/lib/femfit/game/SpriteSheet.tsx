/* eslint-disable react-native/no-inline-styles */
import React from 'react';
import {
  Animated,
  Easing,
  LayoutChangeEvent,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';

import constants from './constants';

const {SPRITE_SIZE} = constants;

type Animation = {
  name: string;
  row: number;
  frames?: number;
};

type InternalAnim = {
  frames: number;
  input: number[];
  output: number[];
  translateY: number;
};

type InternalAnims = {[key: string]: InternalAnim};

export type SpriteSheetProps = {
  src: any;
  rows: number;
  cols: number;
  rate?: number;
  anims: Animation[];
  defaultAnim: string;
  style?: StyleProp<ViewStyle>;
};

type SpriteSheetState = {
  time: Animated.Value;
  iAnims: InternalAnims;
  currAnim: string;
  size: number;
  loaded: boolean;
  playing: boolean;
};

/**
 * Component for rendering an animated sprite sheet
 * @see {@link SpriteSheetProps}
 * @see {@link SpriteSheetState}
 */
export default class SpriteSheet extends React.Component<
  SpriteSheetProps,
  SpriteSheetState
> {
  constructor(props: SpriteSheetProps) {
    super(props);
    const iAnims: InternalAnims = {
      [this.props.defaultAnim]: {
        frames: 0,
        input: [0, 0],
        output: [0, 0],
        translateY: 0,
      },
    };

    this.state = {
      time: new Animated.Value(1.0),
      iAnims,
      currAnim: this.props.defaultAnim,
      size: SPRITE_SIZE,
      loaded: false,
      playing: false,
    };
  }

  play() {
    this.state.time.setValue(0);
    Animated.timing(this.state.time, {
      toValue: this.state.iAnims[this.state.currAnim].frames - 1,
      duration: 1000,
      easing: Easing.linear,
      useNativeDriver: true,
    }).start(({finished}) => {
      if (!finished) {
        return;
      }

      if (!this.state.playing) {
        this.setState((prevState) => ({
          ...prevState,
          playing: true,
        }));
        this.play();
      } else {
        this.play();
      }
    });
  }

  stop() {
    this.state.time.stopAnimation();
  }

  generateInterpolationRanges(event: LayoutChangeEvent) {
    if (this.state.loaded) {
      return;
    }

    const size = event.nativeEvent.layout.height;

    const iAnims: InternalAnims = {};
    for (const anim of this.props.anims) {
      const nFrames = anim.frames ?? this.props.cols;

      const iAnim: InternalAnim = {
        frames: nFrames,
        input: getNumPairs(nFrames).slice(0, 2 * nFrames - 1),
        output: [0].concat(
          getNumPairs(nFrames - 1).map((i) => -(i + 1) * size),
        ),
        translateY: anim.row * size,
      };
      iAnims[anim.name] = iAnim;
    }

    this.setState(
      {
        ...this.state,
        iAnims,
        size,
        loaded: true,
      },
      () => {
        this.play();
      },
    );
  }

  render(): React.ReactNode {
    return (
      <View
        style={[styles.container, this.props.style]}
        onLayout={(event) => this.generateInterpolationRanges(event)}>
        <Animated.Image
          source={this.props.src}
          resizeMode="stretch"
          style={[
            styles.img,
            {
              opacity: this.state.playing ? 1 : 0,
              height: this.state.size,
              width: this.state.size * this.props.cols,
              transform: [
                {
                  translateX: this.state.time.interpolate({
                    inputRange: this.state.iAnims[this.state.currAnim].input,
                    outputRange: this.state.iAnims[this.state.currAnim].output,
                  }),
                },
                {translateY: this.state.iAnims[this.state.currAnim].translateY},
              ],
            },
          ]}
        />
      </View>
    );
  }
}

const getNumPairs = (length: number) => {
  const arr: number[] = [];
  for (let i = 0; i < length; i++) {
    arr.push(i);
    arr.push(i);
  }
  return arr;
};

const styles = StyleSheet.create({
  container: {
    aspectRatio: 1,
    overflow: 'hidden',
  },
  img: {},
});
