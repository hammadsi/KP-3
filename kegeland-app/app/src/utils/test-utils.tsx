/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
