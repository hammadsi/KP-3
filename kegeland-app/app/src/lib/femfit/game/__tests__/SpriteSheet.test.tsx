import React from 'react';

import {render} from '~utils/test-utils';
import PlayerSpriteSheet from '~assets/femfit/bird.png';

import SpriteSheet, {SpriteSheetProps} from '../SpriteSheet';

describe('Test SpriteSheet-component', () => {
  it('should render correctly', () => {
    const props: SpriteSheetProps = {
      src: PlayerSpriteSheet,
      rows: 1,
      cols: 3,
      anims: [{name: 'default', row: 0, frames: 3}],
      defaultAnim: 'default',
    };
    const component = <SpriteSheet {...props} />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
