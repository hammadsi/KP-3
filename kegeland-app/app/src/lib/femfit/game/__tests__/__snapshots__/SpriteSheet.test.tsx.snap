// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SpriteSheet-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "aspectRatio": 1,
          "overflow": "hidden",
        },
        undefined,
      ]
    }
  >
    <Image
      collapsable={false}
      nativeID="animatedComponent"
      resizeMode="stretch"
      source={
        Object {
          "testUri": "../../../assets/femfit/bird.png",
        }
      }
      style={
        Object {
          "height": 128,
          "opacity": 0,
          "transform": Array [
            Object {
              "translateX": 0,
            },
            Object {
              "translateY": 0,
            },
          ],
          "width": 384,
        }
      }
    />
  </View>
</View>
`;
